File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/concrete/ConcreteScreen.java
package com.jaynewstrom.screenswitchersample.concrete;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import com.jaynewstrom.concrete.Concrete;
import com.jaynewstrom.concrete.ConcreteBlock;
import com.jaynewstrom.concrete.ConcreteWall;
import com.jaynewstrom.screenswitcher.Screen;
import com.jaynewstrom.screenswitchersample.MainActivityComponent;
import com.jaynewstrom.screenswitchersample.ScreenSwitcherApplication;

public abstract class ConcreteScreen<C> implements Screen {

    private ConcreteWall<C> screenWall;

    @Override public final View createView(Context context, ViewGroup hostView) {
        ConcreteWall<MainActivityComponent> activityWall = Concrete.findWall(context);
        screenWall = activityWall.stack(block(activityWall.getComponent()));
        return createView(screenWall.createContext(context), screenWall.getComponent());
    }

    protected abstract ConcreteBlock<C> block(MainActivityComponent theParentComponent);

    protected abstract View createView(Context context, C component);

    @Override public final void destroyScreen(View viewToDestroy) {
        ScreenSwitcherApplication.watchObject(viewToDestroy.getContext(), this);
        ScreenSwitcherApplication.watchObject(viewToDestroy.getContext(), viewToDestroy.getContext());
        ScreenSwitcherApplication.watchObject(viewToDestroy.getContext(), screenWall.getComponent());
        screenWall.destroy();
    }
}

File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/concrete/ConcreteScreen.kt
package com.jaynewstrom.screenswitchersample.concrete

import android.content.Context
import android.view.View
import android.view.ViewGroup

import com.jaynewstrom.concrete.Concrete
import com.jaynewstrom.concrete.ConcreteBlock
import com.jaynewstrom.concrete.ConcreteWall
import com.jaynewstrom.screenswitcher.Screen
import com.jaynewstrom.screenswitchersample.MainActivityComponent
import com.jaynewstrom.screenswitchersample.ScreenSwitcherApplication

internal abstract class ConcreteScreen<C> : Screen {
    private lateinit var screenWall: ConcreteWall<C>

    override fun createView(context: Context, hostView: ViewGroup): View {
        val activityWall = Concrete.findWall<ConcreteWall<MainActivityComponent>>(context)
        screenWall = activityWall.stack(block(activityWall.component))
        return createView(screenWall.createContext(context), screenWall.component)
    }

    protected abstract fun block(theParentComponent: MainActivityComponent): ConcreteBlock<C>

    protected abstract fun createView(context: Context, component: C): View

    override fun destroyScreen(viewToDestroy: View) {
        ScreenSwitcherApplication.watchObject(viewToDestroy.context, this)
        ScreenSwitcherApplication.watchObject(viewToDestroy.context, viewToDestroy.context)
        ScreenSwitcherApplication.watchObject(viewToDestroy.context, screenWall.component as Any)
        screenWall.destroy()
    }
}
